### ISSUE TITLE
Security: "poisoning" objcts with additional properties slips through ReflectedClass#matchesValue

### ISSUE THREAD

**Issue by bogeeee** [typescript-rtti/typescript-rtti#92]
`matchesValue` does not report, when objects contain properties which are not specified. Contrary the typescript compiler reports this as an error. I see this as security relevent in my case as my RPC framework [restfuncs](https://www.npmjs.com/package/restfuncs) wants to use this checking for all input from the evil client.  So an attacker could sneak in some properties that will get stored in the database, get evaluated by third party libs etc.. Could you have a look at it ?

Reproduce it with the following jest code:

```typescript
    class User {
        id : number;
    }

    const reflectedClass = reflect(User);

    expect(
        reflectedClass.matchesValue({id: 1, poisonedProperty: true})
    ).toBeFalsy()
```

**Comment by rezonant** [comment]
Agreed. I think we may want to be able to do both kinds of checks. matchesValue has too many positional parameters, so I think we'll change it to an options object and add something like allowExtraProperties. That'll be a new minor version for compat. Not sure which mode should be the default yet.

**Comment by bogeeee** [comment]
Cool, thx ! 
I'd feature the secure version as default. Also that would be the expected behavior in a typescript sense.

**Comment by rezonant** [comment]
It's [not quite that simple](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgILIN4Chm+TAewIC5kQBXAWwCNoBuLAXyywBsIxkwALUAc2QBeTPiKkAjACYAzABpk1OFFLTJ45IwbtOcUumE9+dIA) in Typescript.

Nonetheless, we might go for allowExtraProperties=false as the default just to avoid the footgun.

**Comment by bogeeee** [comment]
Ah lol, you're right. I was confused about that, see that issue.

**Comment by bogeeee** [comment]
Also an (optional) automatic cleaning of extra properties would be very handy in my RPC case. Cause you never know what extra properties might get attached by other client frameworks and we don't want to bother the user with an error then.

So if you're implementing the checks anyway...

**Comment by rezonant** [comment]
Oops this wasn't closed when 0.9.0 shipped. You must now opt in to allowing extra properties as discussed.
